components/brave_news/browser/feed_v2_builder.cc

```diff
@@ -260,31 +286,80 @@ int GetNormal(int min, int max) {
   return min + floor((max - min) * GetNormal());
 }
 
-using GetWeighting = double(const ArticleWeight& weight);
+using GetWeighting = std::function<double(const mojom::FeedItemMetadataPtr& metadata, const ArticleWeight& weight)>;
+
+// Returns a probability distribution (sum to 1) of the weights. Temperature
+// controls how "smooth" the distribution is. High temperature brings the
+// distribution closer to a uniform distribution (more randomness).
+// Low temperature brings the distribution closer to a delta function (less
+// randomness).
+void SoftmaxWithTemperature(std::vector<double>& weights, double temperature) {
+  if (temperature == 0)
+    return;
+
+  double max = *base::ranges::max_element(weights.begin(), weights.end());
+  base::ranges::transform(weights.begin(), weights.end(), weights.begin(),
+                 [temperature, max](double weight) {
+                    return std::exp((weight - max) / temperature);
+                 });
+  double sum = std::accumulate(weights.begin(), weights.end(), 0.0);
+  base::ranges::transform(weights.begin(), weights.end(), weights.begin(),
+                 [sum](double weight) { return weight / sum; });
+}
+
+// Sample across subscribed channels (direct and native) and publishers.
+std::vector<GroupSample> SampleSubscribedGroups(
+    const Channels& channels,
+    const std::vector<std::string>& eligible_content_groups) {
+  std::vector<GroupSample> subscribed_groups_sample;
+
+  if (eligible_content_groups.empty()) {
+    return subscribed_groups_sample;
+  }
+
+  for (int i=0; i < features::kBraveNewsMaxBlockCards.Get() + 1; i++) {
```

Comment:
So we generate at least as many content groups as we have cards in the feed - maybe we should have this return a dynamic number of SubscribedGroups, and then `GenerateBlock` could use the number of content_groups to determine the number of cards? 

Alternatively, we could change the function signature to something like:

`GenerateBlock(articles, inline_discovery_ratio, sample_groups=true/false)` and have the `GenerateBlock` function call `SampleSubscribedGroups` when it needs a new group - that way we wouldn't need to generate all of these up front.

Reply:
tried something in between, as we need some internal state to get the content groups. Let me know what you think